proxy/jsonproto.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"os"
+)
+
+type Request struct {
+	Id      uint             `json:"id"`
+	Command string           `json:"command"`
+	Data    *json.RawMessage `json:"data"`
+}
+
+type Response struct {
+	Id      uint                   `json:"id"`
+	Success bool                   `json:"success"`
+	Error   string                 `json:"error,omitempty"`
+	Data    map[string]interface{} `json:"data,omitempty"`
+}
+
+// XXX: could do with its own package to remove that ugly namespacing
+type JsonProtoHandler func([]byte, interface{}) (map[string]interface{}, error)
+
+type JsonProto struct {
+	handlers map[string]JsonProtoHandler
+}
+
+func NewJsonProto() *JsonProto {
+	return &JsonProto{
+		handlers: make(map[string]JsonProtoHandler),
+	}
+}
+
+func (proto *JsonProto) Handle(cmd string, handler JsonProtoHandler) {
+	proto.handlers[cmd] = handler
+}
+
+type clientCtx struct {
+	conn net.Conn
+
+	reader  *bufio.Reader
+	encoder *json.Encoder
+
+	userData interface{}
+}
+
+func (proto *JsonProto) handleRequest(ctx *clientCtx, req *Request) (*Response, error) {
+	handler, ok := proto.handlers[req.Command]
+	if !ok {
+		return nil, fmt.Errorf("couldn't find command '%s'", req.Command)
+	}
+
+	var resp *Response
+	data := ([]byte)(*req.Data)
+
+	if respMap, err := handler(data, ctx.userData); err != nil {
+		resp = &Response{
+			Id:      req.Id,
+			Success: false,
+			Error:   err.Error(),
+			Data:    respMap,
+		}
+	} else if respMap != nil {
+		resp = &Response{
+			Id:      req.Id,
+			Success: true,
+			Data:    respMap,
+		}
+	} else {
+		resp = &Response{
+			Id:      req.Id,
+			Success: true,
+		}
+	}
+
+	return resp, nil
+}
+
+func (proto *JsonProto) Serve(conn net.Conn, userData interface{}) {
+	ctx := &clientCtx{
+		conn:     conn,
+		userData: userData,
+		reader:   bufio.NewReader(conn),
```

Comment:
I liked the new line separator between requests, but it shouldn't be needed if we just use json.NewDecoder() and it's true that line oriented messages are more CPU work as well (to find the '\n' in the input data). Will change.